primitives/src/merkle_tree/namespaced_merkle_tree/mod.rs

```diff
@@ -221,22 +217,24 @@ where
         &mut self,
         elems: impl IntoIterator<Item = impl core::borrow::Borrow<Self::Element>>,
     ) -> Result<(), PrimitivesError> {
-        // TODO(#314): update namespace metadata
-        self.inner.extend(elems)
+        let leaves =
+            NMT::<E, H, Arity, N, T>::update_namespace_metadata(&mut self.namespace_ranges, elems)?;
+        self.inner.extend(leaves)
     }
 
     fn push(
         &mut self,
         elem: impl core::borrow::Borrow<Self::Element>,
     ) -> Result<(), PrimitivesError> {
-        // TODO(#314): update namespace metadata
-        self.inner.push(elem)
+        let cloned = elem.borrow().clone();
```

Comment:
At first I was confused why we need to clone here but not in `extend`. I wonder if it would be clearer to `let leaves = NMT::update_namespace_metadata(...)` and then `self.inner.push(leaves[0])` or `self.inner.extend(leaves)` (even though `leaves` only has 1 element). To more closely mirror `extend`. Definitely not a big deal either way, though

Reply:
I wonder if it would make sense to implement this as `self.extend([elem])` instead?

Reply:
Yeah that would be quite elegant

Reply:
Great idea